### DISCUSSION TITLE
Fetch Camera Frames on Android

### DISCUSSION THREAD

**Discussion by dickenhobelix** [fyne-io/fyne#5219] [Q&A]
Hi,

I am not entirely sure if this is the right place for this question - if not please just close/delete it.

I am trying to write a qr-code-reader-style android app using fyne, that can display and pipe camera frames into a qr reader library. Is there any way to do this in fyne? All hints I have found so far do not seem to work with android (like using opencv bindings or accessing raw v4l2 devices). Maybe anyone has some hints for me?

Thanks and regards
Dickenhobelix

**Comment by andydotxyz** [comment]
We don't have anything baked in yet - but I know that some people have succeeded doing it in their own apps.
Sadly I don't have any open source projects to point you to! If we can make this work within the project we will all be happy :).

**Reply by dickenhobelix** [reply]
Hi, thanks for your reply. Do you think that invoking the Camera API in the Android SDK via the NDK from fine could be feasible?

**Reply by andydotxyz** [reply]
Personally I haven't tried - but it technically should be possible. We do plenty of SDK access already.

**Reply by dickenhobelix** [reply]
Thanks, I am going to look into this, maybe I can reactivate some of my back-in-the-day java skills...

Do you have any particular good example in the code you recommend to look into for SDK access?

**Comment by hkparker** [comment]
I've been looking into how to do this myself if possible.  It's not important enough to me yet to really make a push for it, but I'll share everything I've found on the topic here:

1. Firing off an intent can be done via the JNI and I have a working example of this [here](https://github.com/fyne-io/fyne/issues/2790#issuecomment-2430375023)
2. Receiving the intent responses requires modifying Fyne's `GoNativeActivity.java` around [here](https://github.com/fyne-io/fyne/blob/master/internal/driver/mobile/app/GoNativeActivity.java#L305) to add the responses you care about
3. Fyne will need to declare dependencies on `androidx.camera.[...]` for camera and `com.google.mlkit.[...]` for barcode scanning
4. You'll need to use a custom `AndroidManifest.xml` to declare camera permissions

Take a look at:
* This [example app](https://github.com/fynelabs/devicename) (and related [blog post](https://fynelabs.com/2024/03/01/running-native-android-code-in-a-fyne-app/))
* This [previous discussion on the topic](https://github.com/fyne-io/fyne/discussions/4965)

For modifying that Java file, you'll also need to communicate the results back to go of course, and looking around at how the file picker stuff works, that's going to involve going back through C.  Search around for the symbols used in that process to get an idea of how that works.

**Reply by dickenhobelix** [reply]
Thank you so much, that helps a lot. I try to get something up and running, maybe I can even manage to contribute something back upstream once I did so.

Thank you for sharing!
